src/uct/tcp/tcp_ep.c

```diff
@@ -2067,17 +2073,28 @@ ucs_status_t uct_tcp_ep_flush(uct_ep_h tl_ep, unsigned flags,
     }
 
     status = uct_tcp_ep_check_tx_res(ep);
-    if (status == UCS_ERR_NO_RESOURCE) {
-        UCT_TL_EP_STAT_FLUSH_WAIT(&ep->super);
-        return UCS_ERR_NO_RESOURCE;
+    if (ucs_unlikely(status != UCS_OK)) {
+        if (status == UCS_ERR_NO_RESOURCE) {
+            return UCS_ERR_NO_RESOURCE;
+        }
+        return UCS_OK;
     }
 
-    if (ep->flags & UCT_TCP_EP_FLAG_PUT_TX_WAITING_ACK) {
-        status = uct_tcp_ep_put_comp_add(ep, comp, ep->tx.put_sn);
+    if (ep->last_acked_sn != ep->tx.sn) {
+        /* Decrement the sequence number to not consider the flush operation
+         * for waiting ACK, the sequence number will be incremented in PUT
+         * Zcopy operation. PUT Zcopy sends PUT REQ message which triggers
+         * sending ACK message back. */
+        --ep->tx.sn;
+        status = uct_ep_put_zcopy(&ep->super.super, NULL, 0, 0, 0, NULL);
```

Comment:
why shutdown doesn't work?

Reply:
`shutdown()` doesn’t work if user ha star following flow:
```
ucp_ep_close_nbx(FLUSH);
ucp_worker_destroy();
```
Since `uct_ep_close()` shutdowns the connection and we should wait for it completion in `epoll_wait()`. So, real closing of the socket is deferred. And we destroy socket, when destroying Worker. It will be ok, if user had some the following flow:
```
ucp_ep_close_nbx(FLUSH);
// wait for some time to ensure that socket is closed after sholutdown
ucp_worker_destroy();

```